pkg/deploy/stack_deployer.go

```diff
@@ -54,35 +60,140 @@ type latticeServiceStackDeployer struct {
 	svcBuilder            gateway.LatticeServiceBuilder
 }
 
+var tgGcOnce sync.Once
+var tgGc *TgGc
+
 func NewLatticeServiceStackDeploy(
 	log gwlog.Logger,
 	cloud pkg_aws.Cloud,
 	k8sClient client.Client,
 ) *latticeServiceStackDeployer {
 	brTgBuilder := gateway.NewBackendRefTargetGroupBuilder(log, k8sClient)
 
+	tgMgr := lattice.NewTargetGroupManager(log, cloud)
+	tgSvcExpBuilder := gateway.NewSvcExportTargetGroupBuilder(log, k8sClient)
+	svcBuilder := gateway.NewLatticeServiceBuilder(log, k8sClient, brTgBuilder)
+
+	tgGcOnce.Do(func() {
+		// TODO: need to refactor TG synthesizer. Remove stack from constructor
+		// arguments and use it as Synth argument. That will help with Synth
+		// reuse for GC purposes
+		tgGcSynth := lattice.NewTargetGroupSynthesizer(log, cloud, k8sClient, tgMgr, tgSvcExpBuilder, svcBuilder, nil)
+		tgGcFn := NewTgGcFn(tgGcSynth)
+		tgGc = &TgGc{
+			lock:    sync.Mutex{},
+			log:     log.Named("gc"),
+			ctx:     context.TODO(),
+			isDone:  atomic.Bool{},
+			ivl:     TG_GC_IVL,
+			cycleFn: tgGcFn,
+		}
+		tgGc.start()
+	})
+
 	return &latticeServiceStackDeployer{
 		log:                   log,
 		cloud:                 cloud,
 		k8sClient:             k8sClient,
 		latticeServiceManager: lattice.NewServiceManager(log, cloud),
-		targetGroupManager:    lattice.NewTargetGroupManager(log, cloud),
+		targetGroupManager:    tgMgr,
 		targetsManager:        lattice.NewTargetsManager(log, cloud),
 		listenerManager:       lattice.NewListenerManager(log, cloud),
 		ruleManager:           lattice.NewRuleManager(log, cloud),
 		dnsEndpointManager:    externaldns.NewDnsEndpointManager(log, k8sClient),
-		svcExportTgBuilder:    gateway.NewSvcExportTargetGroupBuilder(log, k8sClient),
-		svcBuilder:            gateway.NewLatticeServiceBuilder(log, k8sClient, brTgBuilder),
+		svcExportTgBuilder:    tgSvcExpBuilder,
+		svcBuilder:            svcBuilder,
 	}
 }
 
+type TgGcCycleFn = func(context.Context) (TgGcResult, error)
+
+func NewTgGcFn(tgSynth *lattice.TargetGroupSynthesizer) TgGcCycleFn {
+	return func(ctx context.Context) (TgGcResult, error) {
+		t0 := time.Now()
+		results, err := tgSynth.SynthesizeUnusedDelete(ctx)
+		if err != nil {
+			return TgGcResult{}, err
+		}
+		succ := 0
+		for _, res := range results {
+			if res.Err == nil {
```

Comment:
Do we need to print `res.Err` or do errors.Join(res.Err) and return it with TgGcResult{}?

Reply:
Not really, mostly errors happen when TG is in state that cannot be deleted yet. It pollutes logs.

Reply:
I think critical errors needs to be logged in SynthesizeUnusedDelete, or need to wrap them into some "NonOk" types and do type check here.